include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,301 @@
+// Copyright (C) 2020 - AnastaZIuk
+// This file is part of the "IrrlichtBAW" engine.
+// For conditions of distribution and use, see copyright notice in irrlicht.h
+
+#ifndef __IRR_C_SUMMED_AREA_TABLE_IMAGE_FILTER_H_INCLUDED__
+#define __IRR_C_SUMMED_AREA_TABLE_IMAGE_FILTER_H_INCLUDED__
+
+#include "irr/core/core.h"
+
+#include <type_traits>
+#include <functional>
+
+#include "irr/asset/filters/CMatchedSizeInOutImageFilterCommon.h"
+#include "CConvertFormatImageFilter.h"
+
+namespace irr
+{
+namespace asset
+{
+
+template<bool ExclusiveMode>
+class CSummedAreaTableImageFilterBase
+{
+	public:
+		class CSummStateBase
+		{
+			public:
+				
+				static inline constexpr size_t decodeTypeByteSize = sizeof(double);
+				uint8_t*	scratchMemory = nullptr;										//!< memory covering all regions used for temporary filling within computation of sum values
+				size_t	scratchMemoryByteSize = {};											//!< required byte size for entire scratch memory
+				bool normalizeImageByTotalSATValues = false;								//!< after sum performation, the option decide whether to divide the entire image by the max sum values in (maxX, 0, z) depending on input image - no divison will be performed if an input is UNORM.
+				
+				static inline size_t getRequiredScratchByteSize(const ICPUImage* inputImage, const ICPUImage* outputImage)
+				{
+					const auto& inputCreationParams = inputImage->getCreationParameters();
+					const auto channels = asset::getFormatChannelCount(inputCreationParams.format);
+					const auto regions = inputImage->getRegions();
+					const auto& trueExtent = inputCreationParams.extent;
+
+					size_t retval = cachesOffset = trueExtent.width * trueExtent.height * trueExtent.depth * inputCreationParams.arrayLayers * channels * decodeTypeByteSize;
+
+					retval += (imageTotalScratchCacheByteSize = channels * decodeTypeByteSize * inputImage->getCreationParameters().extent.depth * inputCreationParams.arrayLayers);
```

Comment:
you do realize that you can save on scratch memory by only processing "one layer at a time" because you do not sum across layers?

plus there's no need for caches outside of your decode scratch space

Reply:
yeah yeah, got it. Actually I didn't figure it out, my vision was different